### PR TITLE
JENKINS-45643 Fixing bug with inifinite running jobs.

### PR DESCRIPTION
**Pull request by benjaminfuchs**
We are using this already in our productive system as a workraround for the issue JENKINS-45643.
Details can be found in the comment:
https://issues.jenkins-ci.org/browse/JENKINS-45643?focusedCommentId=334416&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-334416

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java [hunk 1]
```diff
@@ -143,11 +155,20 @@ private static void createStatus(@NonNull Run<?, ?> build, @NonNull TaskListener
 
     private static void sendNotifications(Run<?, ?> build, TaskListener listener)
             throws IOException, InterruptedException {
-        final SCMSource s = SCMSource.SourceByItem.findSource(build.getParent());
+        SCMSource s = SCMSource.SourceByItem.findSource(build.getParent());
```
Comment (benjaminfuchs) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java]: This line will fail if the job already got "disabled" (e.g. due to not building feature branches that are also build in a Pull Request).
The job will forever be stuck in Bitbucket as "In Progress" and the user will never be able to merge it.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java [hunk 2]
```diff
@@ -143,11 +155,20 @@ private static void createStatus(@NonNull Run<?, ?> build, @NonNull TaskListener
...
+
         if (!(s instanceof BitbucketSCMSource)) {
-            return;
+            if (sourceCache.containsKey(buildName)) {
+                s = sourceCache.get(buildName);
+            } else {
+                return;
+            }
+        } else {
+            sourceCache.put(buildName, s);
```
Comment (benjaminfuchs) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketBuildStatusNotifications.java]: Remember the last 1000 SCMSources so we can still update the status to Bitbucket.

**Comment by casz** [comment]
Not sure understand how a cache solves JENKINS-45643 🤔

**Comment by benjaminfuchs** [comment]
@casz The cache helps updating job that already got disabled and "findSource()" will return null. I added it as a workaround since we have been  blocked by this ("running" feature branch jobs makes it impossible to merge a PR). I tried to describe it in more detail in the jira ticket.

**Comment by casz** [comment]
I think the right fix is #147 rather than this :)

**Comment by benjaminfuchs** [comment]
@casz #147 looks similar thanks for the hint.

**Comment by benjaminfuchs** [comment]
@casz seems #147 was reverted since it has some side effects. This PR fixes the issue without side effects.

**Comment by mrkita** [comment]
benjamin, I just tried cloning and testing a build of this branch:
https://github.com/benjaminfuchs/bitbucket-branch-source-plugin/commits/JENKINS-45643

and it did not work for me. The infinite job wait was still occuring! Is that version different than this pull request, or does something else need to be configured?

If this fix is not desirable to the maintainers (which I can understand), can we please get a proper fix? I see that a fix was reverted due to the job counts reported, but now we are in some broken noman's land.

I tried out 2.2.13 and it works as expected. I will run that for now for our team, but I can't say thats anything but a short term bridge, considering this plugin is a keystone for our operations.

**Pull request closed without merging**